### PR TITLE
Reuse AFHTTPSessionManagers if possible to avoid race condition crash

### PR DESCRIPTION
**Pull request by lehrblogger**
tl;dr: Using multiple `AFHTTPSessionManager`s for a given baseURL can result in a race condition involving overreleased `SecTrustRef`s that crash the application. This change attempts to re-use those managers as much as possible, and will hopefully fix the crash.

-----

It seems that the `NSURLAuthenticationChallenge` objects passed to `NSURLSessionDelegate`’s `URLSession:didReceiveChallenge:completionHandler:` share their `protectionSpace.serverTrust SecTrustRef`s. This is perhaps a bug in `NSURLSession`, but if they do share `SecTrustRef`s, then there's [an old Apple Listserv post](http://lists.apple.com/archives/apple-cdsa/2010/Apr/msg00017.html) that suggests it may cause a certain class of crashes related to `CFNetworkTrust::evaluate()`.

By creating a new `AFHTTPSessionManager` every time a request is made to a particular `baseURL`, there might sometimes be multiple concurrent callbacks to `NSURLSessionDelegate`'s `URLSession:didReceiveChallenge:completionHandler`, which the managers implement. From there, the `challenge.protectionSpace.serverTrust` [is passed](https://github.com/AFNetworking/AFNetworking/blob/3.1.0/AFNetworking/AFURLSessionManager.m#L978) to its `AFSecurityPolicy` when applicable, and then later, the SecTrustRef is evaluated and eventually released ([here](https://github.com/AFNetworking/AFNetworking/blob/3.1.0/AFNetworking/AFSecurityPolicy.m#L74) or [here](https://github.com/AFNetworking/AFNetworking/blob/3.1.0/AFNetworking/AFSecurityPolicy.m#L135)). If a `SecTrustRef` is shared between requests to a given `baseURL`, then there could be a race condition in which multiple simultaneous requests overrelease the `SecTrustRef` and a subsequent attempt to use it crashes the app.

-----

Note that the crash may still occur when using multiple `AFHTTPSessionManager`s with different `requestType`s and `responseType`s for a given `baseURL`. That seems significantly less likely to happen, however, so I wanted to see if this fixes the issue before attempting more significant changes.

It has to be in an `@synchronized` block because `NSMutableDictionary` is not thread safe, and because we don't want multiple managers created for a single key.

This may eventually be fixed by https://github.com/AFNetworking/AFNetworking/pull/3616, but this seems like a good change for `SGHTTPRequest` as well.

### CODE REVIEW

**Comment by lehrblogger** [comment]
Ah ha! _This_ is why I had deja vu while I was working on this: https://github.com/seatgeek/SGHTTPRequest/commit/b0745c38f4ab56fd03cf1cfb81ae8b397f976490.

I can't reproduce the original issue that prompted me to open https://github.com/seatgeek/SGHTTPRequest/pull/2, but fortunately I now see a solution that grants the best of both worlds: keep the manager, and recreate only the `requestSerializer` and `responseSerializer`.

**Comment by lehrblogger** [comment]
hmm, do we need to worry about one request overwriting the headers set by a concurrent request using the same manager?

**Comment by lehrblogger** [comment]
Also, I think the builds are failing because [Travis is having issues](https://www.traviscistatus.com/incidents/b4xk2bybjc9d) – we can retry tomorrow.

**Comment by staminajim** [comment]
From the log travis is failing on the Podfile in the test project - maybe the syntax is disagreeable with the travis setup.

> keep the manager, and recreate only the requestSerializer and responseSerializer.

At first glance it _looks_ like this'll work.  Looks like when you start the session manager operation it converts the `requestSerializer` into the `NSMutableURLRequest` object at the request start, and then doesn't use it again until next time you initiate one.

I reckon get this in early, confirm that the request headers and responses aren't shared for our requests, and we'll soon find out if there's anything wrong with it (the whole app will break if it doesn't :D ).

**Comment by lehrblogger** [comment]
@staminajim Thanks, that's good to know about what happens when the request starts. (Whoops, sorry for the mention @synchronize!)

I continued to worry about race conditions where simultaneous requests would mess with each other's serializers, since these might only happen in certain special cases and would be tricky to properly diagnose. I came up with the above fix, which moves most everything in the `start` method into a single `@synchronize` block. Every time there's a new request, the serializers are recreated and configured as necessary. This has the added benefit of simplifying the manager cacheing, which ensures that there is only one manager making a request to a given baseURL at any time – I was worried about this above.

The split diff, or the [diff ignoring whitespace](https://github.com/seatgeek/SGHTTPRequest/pull/6/files?w=1), make it a little easier to see what's going on. If this last change looks ok, let's see how this feels in the app. I'm a little concerned about performance issues with `@synchronize`, but I'm not sure if there's a better way to fix the crash.

**Comment by staminajim** [comment]
>  continued to worry about race conditions where simultaneous requests would mess with each other's serializers, since these might only happen in certain special cases and would be tricky to properly diagnose.

I dont think this is a problem since we only support initiating requests on the main thread.  The request serializer parameters are captured at that point and passed off to the NSOperation objct for use in the background.

Either way - lets get it in and being used.
